### PR TITLE
feat(tiles): execute traversal optimization

### PR DESCRIPTION
**Pull request by belom88**

### CODE REVIEW

File: modules/tiles/src/tileset/traversers/i3s-frame-counter.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+export default class FrameCounter {
+  private frameNumberMap: Map<number, number> = new Map();
+
+  increase(frameNumber: number) {
+    const oldCount = this.frameNumberMap.get(frameNumber) || 0;
+    this.frameNumberMap.set(frameNumber, (oldCount || 0) + 1);
+  }
+
+  decrease(frameNumber: number) {
```
Comment (ibgreen) [inline modules/tiles/src/tileset/traversers/i3s-frame-counter.ts:13]: FrameCounter sounds like it measures rendered frames. This is something else right, measuring pending tiles?

Also I would typically expect something called a frame counter to be "monotonic"? In what case can it decrease?

File: modules/tiles/src/tileset/traversers/i3s-tile-manager.ts [hunk 2]
```diff
@@ -7,33 +10,44 @@ const STATUS = {
...
-  add(request, key, callback, frameState) {
+  add(request, key, callback, frameState: FrameState) {
     if (!this._statusMap[key]) {
+      const {frameNumber} = frameState;
       this._statusMap[key] = {request, callback, key, frameState, status: STATUS.REQUESTED};
+      this.frameCounter.increase(frameNumber);
       request()
         .then((data) => {
           this._statusMap[key].status = STATUS.COMPLETED;
+          this.frameCounter.decrease(frameNumber);
```
Comment (ibgreen) [inline modules/tiles/src/tileset/traversers/i3s-tile-manager.ts:27]: Please add comments, what do we gain by adding arbitrary frame numbers to a counter (as opposed to just incremnting and decrementing it?

File: modules/tiles/src/tileset/traversers/i3s-tile-manager.ts [hunk 3]
```diff
@@ -7,33 +10,44 @@ const STATUS = {
...
+      this.frameCounter.increase(frameState.frameNumber);
       this._statusMap[key].frameState = frameState;
     }
   }
 
   find(key) {
     return this._statusMap[key];
   }
+
+  hasWithFrameState(frameNumber): boolean {
```
Comment (ibgreen) [inline modules/tiles/src/tileset/traversers/i3s-tile-manager.ts:50]: hasPendingTiles?

File: modules/tiles/src/tileset/traversers/i3s-frame-counter.ts (hunk 1 above)
Comment (belom88) [reply modules/tiles/src/tileset/traversers/i3s-frame-counter.ts:13]: It decrease if tile is loaded or error occurred

Comment (belom88) [reply modules/tiles/src/tileset/traversers/i3s-frame-counter.ts:13]: This class accounts how many tiles are in loading progress for particular frameNumber

**Comment by belom88** [comment]
I changed naming

**Review by ibgreen — APPROVED** [review]

**Pull request merged**